firebase-firestore/src/test/java/com/google/firebase/firestore/model/MutationSquashTest.java

```diff
@@ -0,0 +1,449 @@
+    Mutation patch = patchMutation("collection/key", map("foo.bar", "new-bar-value"));
+
+    squashRoundTrips(doc, delete, patch);
+  }
+
+  @Test
+  public void testSquashWithDeleteThenMerge() {
+    MutableDocument doc = doc("collection/key", 0, map("foo", 1));
+    Mutation delete = new DeleteMutation(key("collection/key"), Precondition.NONE);
+    Mutation patch =
+        mergeMutation(
+            "collection/key", map("foo.bar", "new-bar-value"), Arrays.asList(field("foo.bar")));
+
+    squashRoundTrips(doc, delete, patch);
+  }
+
+  @Test
+  public void testSquashPatchThenPatchToDeleteField() {
+    MutableDocument doc = doc("collection/key", 0, map("foo", 1));
+    Mutation patch =
+        patchMutation(
+            "collection/key", map("foo", "foo-patched-value", "bar.baz", FieldValue.increment(1)));
+    Mutation patchToDeleteField =
+        patchMutation(
+            "collection/key", map("foo", "foo-patched-value", "bar.baz", FieldValue.delete()));
+
+    squashRoundTrips(doc, patch, patchToDeleteField);
+  }
+
+  @Test
+  public void testSquashPatchThenMerge() {
+    MutableDocument doc = doc("collection/key", 0, map("foo", 1));
+    Mutation patch =
+        patchMutation(
+            "collection/key", map("foo", "foo-patched-value", "bar.baz", FieldValue.increment(1)));
+    Mutation merge =
+        mergeMutation(
+            "collection/key",
+            map("arrays", FieldValue.arrayUnion(1, 2, 3)),
+            Arrays.asList(field("arrays")));
+
+    squashRoundTrips(doc, patch, merge);
+  }
+
+  @Test
+  public void testSquashArrayUnionThenRemove() {
+    MutableDocument doc = doc("collection/key", 0, map("foo", 1));
+    Mutation union =
+        mergeMutation(
+            "collection/key", map("arrays", FieldValue.arrayUnion(1, 2, 3)), Arrays.asList());
+    Mutation remove =
+        mergeMutation(
+            "collection/key",
+            map("foo", "xxx", "arrays", FieldValue.arrayRemove(2)),
+            Arrays.asList(field("foo")));
+
+    squashRoundTrips(doc, union, remove);
+  }
+
+  @Test
+  public void testSquashSetThenIncrement() {
+    MutableDocument doc = doc("collection/key", 0, map("foo", 1));
+    Mutation set = setMutation("collection/key", map("foo", 2));
+    Mutation update = patchMutation("collection/key", map("foo", FieldValue.increment(2)));
+
+    squashRoundTrips(doc, set, update);
+  }
+
+  @Test
+  public void testSquashSetThenPatchOnDeletedDoc() {
+    MutableDocument doc = deletedDoc("collection/key", 0);
+    Mutation set = setMutation("collection/key", map("bar", "bar-value"));
+    Mutation patch =
+        patchMutation(
+            "collection/key",
+            map("foo", "foo-patched-value", "bar.baz", FieldValue.serverTimestamp()));
+
+    squashRoundTrips(doc, set, patch);
+  }
+
+  @Test
+  public void testSquashFieldDeletionOfNestedField() {
+    MutableDocument doc = doc("collection/key", 0, map("bar.baz", 1));
```

Comment:
This does not created a nested object. Rather it creates an object with a single key of `bar.baz`.